util/opentelemetry-util-genai/src/opentelemetry/util/genai/metrics.py

```diff
@@ -0,0 +1,125 @@
+"""Helpers for emitting GenAI metrics from LLM invocations."""
+
+from __future__ import annotations
+
+import time
+from numbers import Number
+from typing import Dict, Optional
+
+from opentelemetry.metrics import Histogram, Meter
+from opentelemetry.semconv._incubating.attributes import (
+    gen_ai_attributes as GenAI,
+)
+from opentelemetry.trace import Span, set_span_in_context
+from opentelemetry.util.genai.instruments import Instruments
+from opentelemetry.util.genai.types import LLMInvocation
+from opentelemetry.util.types import AttributeValue
+
+_NS_PER_SECOND = 1_000_000_000
+
+
+def _now_ns() -> int:
+    return time.time_ns()
+
+
+def _get_span_start_time_ns(span: Optional[Span]) -> Optional[int]:
+    if span is None:
+        return None
+    for attr in ("start_time", "_start_time"):
+        value = getattr(span, attr, None)
+        if isinstance(value, int):
+            return value
+    return None
+
+
+def _calculate_duration_seconds(span: Optional[Span]) -> Optional[float]:
+    """Calculate duration in seconds from span start time to now."""
+    start_time_ns = _get_span_start_time_ns(span)
+    if start_time_ns is None:
+        return None
+    elapsed_ns = max(_now_ns() - start_time_ns, 0)
+    return elapsed_ns / _NS_PER_SECOND
+
+
+class InvocationMetricsRecorder:
+    """Records duration and token usage histograms for GenAI invocations."""
+
+    def __init__(self, meter: Meter):
+        instruments = Instruments(meter)
+        self._duration_histogram: Histogram = (
+            instruments.operation_duration_histogram
+        )
+        self._token_histogram: Histogram = instruments.token_usage_histogram
+
+    def record(
+        self,
+        span: Optional[Span],
+        invocation: LLMInvocation,
+        *,
+        error_type: Optional[str] = None,
+    ) -> None:
+        """Record duration and token metrics for an invocation if possible."""
+        if span is None:
+            return
+
+        tokens: list[tuple[int, str]] = []
```

Comment:
can you use a more descriptive variable name, I thought this was holding the actual tokens at first.